providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+        G(v1, v6, v11, v12);                                                   \
+        G(v2, v7, v8, v13);                                                    \
+        G(v3, v4, v9, v14);                                                    \
+    } while ((void)0, 0)
+
+# undef BLAKE2_ROUND_COLUMN
+# define BLAKE2_ROUND_COLUMN(x, i)                                             \
+    do {                                                                       \
+        BLAKE2_ROUND(                                                          \
+            x[16 * i],      x[16 * i + 1],  x[16 * i + 2],  x[16 * i + 3],     \
+            x[16 * i + 4],  x[16 * i + 5],  x[16 * i + 6],  x[16 * i + 7],     \
+            x[16 * i + 8],  x[16 * i + 9],  x[16 * i + 10], x[16 * i + 11],    \
+            x[16 * i + 12], x[16 * i + 13], x[16 * i + 14], x[16 * i + 15]     \
+        );                                                                     \
+    } while ((void)0, 0)
+
+# undef BLAKE2_ROUND_ROW
+# define BLAKE2_ROUND_ROW(x, i)                                                \
+    do {                                                                       \
+        BLAKE2_ROUND(                                                          \
+            x[2 * i],      x[2 * i + 1],  x[2 * i + 16],  x[2 * i + 17],       \
+            x[2 * i + 32], x[2 * i + 33], x[2 * i + 48],  x[2 * i + 49],       \
+            x[2 * i + 64], x[2 * i + 65], x[2 * i + 80],  x[2 * i + 81],       \
+            x[2 * i + 96], x[2 * i + 97], x[2 * i + 112], x[2 * i + 113]       \
+        );                                                                     \
+    } while ((void)0, 0)
+
+typedef struct {
+    uint64_t v[ARGON2_QWORDS_IN_BLOCK];
+} block;
+
+typedef enum {
+    ARGON2_VERSION_10 = 0x10,
+    ARGON2_VERSION_13 = 0x13,
+    ARGON2_VERSION_NUMBER = ARGON2_VERSION_13
+} ARGON2_VERSION;
+
+typedef enum {
+    ARGON2_D  = 0,
+    ARGON2_I  = 1,
+    ARGON2_ID = 2
+} ARGON2_TYPE;
+
+typedef struct {
+    uint8_t *out;
+    uint32_t outlen;
+    uint8_t *pwd;
+    uint32_t pwdlen;
+    uint8_t *salt;
+    uint32_t saltlen;
+    uint8_t *secret;
+    uint32_t secretlen;
+    uint8_t *ad;
+    uint32_t adlen;
+    uint32_t t_cost;
+    uint32_t m_cost;
+    uint32_t lanes;
+    uint32_t threads;
+    uint32_t version;
+    uint32_t flags;
+    ARGON2_TYPE type;
+    block *memory;
+    uint32_t passes;
+    uint32_t memory_blocks;
+    uint32_t segment_length;
+    uint32_t lane_length;
+    OPENSSL_CTX *openssl_ctx;
+} ARGON2_CTX;
+
+typedef struct {
+    uint32_t pass;
+    uint32_t lane;
+    uint8_t slice;
+    uint32_t index;
+} ARGON2_POS;
+
+typedef struct {
+    ARGON2_POS pos;
+    ARGON2_CTX *ctx;
+} ARGON2_THREAD_DATA;
+
+union endianity_test {
+    long one;
+    char little;
+} is_endian = { 1 };
+
+typedef struct {
+    void *provctx;
+    ARGON2_CTX ctx;
+} KDF_ARGON2;
+
+static OSSL_OP_kdf_newctx_fn kdf_argon2i_new;
```

Comment:
When you rebase off master these names will have to change to OSSL_FUNC_kdf_xxx_fn

Reply:
Renamed, thanks!